### PR TITLE
Changing the public API - Adding tests + cleanup

### PR DESCRIPTION
**Pull request by Hutch4**
## Summary 

Before, we had an API that just gave you a set of roots, designed to give you the a list of 2 roots, the primary and the secondary storage. 
It didn't consider which type of storage root you wanted, the system or the application directory, and also it assumed you knew the 1st was the primary, the 2nd was the secondary (No typing).

Now we have a new API, 4 methods, get primary/secondary for base/application roots. 
The primary returning a single directory, the secondary returning a list, as it can be 0-* roots (this covers the case we somehow find 2 SD cards, this is a possibility) 

This PR covers:
- changes the API
- refactoring the Inspectors to deal with the new scenario and splitting to call base or application path.
- Adding unit tests to the inspectors.
- General refactoring 

## Testing

We have Unit tests on the Overall inspector, and the individual inspector instances, 
Some of the tests seem very similar across the overall inspector and the individual instances, but I think its good to keep them as if you change the way the general one acts, you still have the tests of the implementations of the instances of each inspector type.

TL;DR we have tests, its good.

## Notes
There are a few points that still need to be neatened up, but I didn't want to keep adding to this branch as its already a big PR. 
I'll try to comment on the sections that need work

### CODE REVIEW

File: demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.java [hunk 1]
```diff
@@ -24,20 +25,20 @@ protected void onCreate(Bundle savedInstanceState) {
         assetCloner = DemoDependenciesFactory.createAssetCloner(getApplicationContext());
 
         AndroidDeviceStorageInspector storageInspector = DemoDependenciesFactory.createStorageInspector(getApplicationContext());
-        List<DeviceStorageRoot> deviceStorageRoots = storageInspector.getDeviceStorageRoots();
+        List<StoragePath> deviceStoragePaths = Collections.emptyList(); // TODO use the inspector :D
```
Comment (Hutch4) [inline demo/src/main/java/com/novoda/storagepathfinder/demo/LandingActivity.java]: This needs to be fixed. We need new logic for what the demo wants to do, as before it was just a list of paths, now there is a choice of which paths we want

File: library/src/main/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspector.java [hunk 2]
```diff
@@ -28,63 +29,78 @@ public AndroidDeviceStorageInspector(
...
-    private String getPrimaryStoragePath() {
-        return fileSystem.getCommonDirectories().getExternalStorageDirectory().getPath();
+    private Set<StoragePath> findActiveSecondaryStorageBasePaths() {
+        return filterToOnlyRootsThatExist(findAllSecondaryStorageBasePaths());
     }
 
-    private String getDirectoryPathAboveTheAndroidFolderFrom(String path) {
-        String basePath = path.split("/.?Android/")[0];
-        return basePath == null ? "" : basePath;
+    private Set<StoragePath> findAllSecondaryStorageBasePaths() {
```
Comment (Hutch4) [inline library/src/main/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspector.java:67]: There is a bit of duplication here, this should be neatened up (potentially)

File: library/src/main/java/com/novoda/storagepathfinder/ExternalFileDirectoryInspector.java [hunk 3]
```diff
@@ -14,8 +14,14 @@
  */
 public class ExternalFileDirectoryInspector implements SecondaryDeviceStorageInspector {
 
+    private static final StoragePath.Type SECONDARY = StoragePath.Type.SECONDARY;
     private static final String ANDROID_PATTERN = "/(Android)";
 
+    public enum Filter {   // TODO kill this soon
```
Comment (Hutch4) [inline library/src/main/java/com/novoda/storagepathfinder/ExternalFileDirectoryInspector.java:20]: I'd like to refactor this class so we can make better use of the repeated logic, and not have to pass down this enum through multiple methods (that don't all use it)

File: library/build.gradle [hunk 4]
```diff
@@ -56,9 +56,11 @@ android {
 dependencies {
     implementation fileTree(include: ['*.jar'], dir: 'libs')
     implementation 'com.android.support:appcompat-v7:27.1.1'
+
     testImplementation 'junit:junit:4.12'
     testImplementation 'org.mockito:mockito-core:2.19.0'
     testImplementation 'com.google.truth:truth:0.41'
+    testImplementation 'org.assertj:assertj-core:1.7.1'
```
Comment (MechaRyRy) [inline library/build.gradle]: What advantage does this have over the `truth` library that was already added?

File: library/src/main/java/com/novoda/storagepathfinder/AndroidCommonDirectories.java [hunk 5]
```diff
@@ -12,23 +14,19 @@
  */
 public class AndroidCommonDirectories implements CommonDirectories {
```
Comment (MechaRyRy) [inline library/src/main/java/com/novoda/storagepathfinder/AndroidCommonDirectories.java]: `CommonDirectories` should probably be changed now, it's referring to the `external` storage directories now.

File: library/src/main/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspector.java [hunk 6]
```diff
@@ -28,63 +29,78 @@ public AndroidDeviceStorageInspector(
...
+
+    @Override
+    public StoragePath getPrimaryStorageApplicationPath() {
+        return primaryStorageInspector.getPrimaryDeviceStorageApplicationPath();
+    }
 
-        Set<DeviceStorageRoot> secondaryStorageRoots = findActiveSecondaryStorageRoots();
+    @Override
+    public List<StoragePath> getSecondaryStorageBasePath() {
+        List<StoragePath> storageRoots = new ArrayList<>(2);
```
Comment (MechaRyRy) [inline library/src/main/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspector.java]: This should use `secondaryStorageInspectors.size` no? Otherwise if you add another to the list this will throw.

File: library/src/main/java/com/novoda/storagepathfinder/DeviceStorageInspector.java [hunk 7]
```diff
@@ -4,5 +4,28 @@
 
 public interface DeviceStorageInspector {
 
-    List<DeviceStorageRoot> getDeviceStorageRoots();
+    /**
+     * The system storage path of your primary storage
```
Comment (MechaRyRy) [inline library/src/main/java/com/novoda/storagepathfinder/DeviceStorageInspector.java]: Fullstop. Keep the javadocs consistent across classes.

Comment (MechaRyRy) [reply library/src/main/java/com/novoda/storagepathfinder/DeviceStorageInspector.java]: Same for the rest in this class.

File: library/src/main/java/com/novoda/storagepathfinder/DeviceStoragePath.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        this.type = type;
+    }
+
+    @Override
+    public Type getType() {
+        return type;
+    }
+
+    @Override
+    public String asString() {
```
Comment (MechaRyRy) [inline library/src/main/java/com/novoda/storagepathfinder/DeviceStoragePath.java]: `asString` implies you are getting the String of both the type and the path, this isn't the case. Opt for being more specific in this case.

File: library/src/main/java/com/novoda/storagepathfinder/DeviceStoragePath.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        return path;
+    }
+
+    @Override
+    public File asFile() {
```
Comment (MechaRyRy) [inline library/src/main/java/com/novoda/storagepathfinder/DeviceStoragePath.java]: `pathAsFile`.

File: library/src/main/java/com/novoda/storagepathfinder/ExternalFileDirectoryInspector.java [hunk 10]
```diff
@@ -27,16 +33,22 @@
...
+    }
+
+    @Override
+    public List<StoragePath> getSecondaryDeviceStorageApplicationPaths() {
         return deviceFeatures.canReportExternalFileDirectories()
-                ? checkExternalFileDirectoriesForRemovableStorage() : Collections.emptyList();
+                ? checkExternalFileDirectoriesForRemovableStorage(Filter.APPLICATION) : Collections.emptyList();
     }
 
     @TargetApi(Build.VERSION_CODES.KITKAT)    // NO LUCK API 16-18
```
Comment (MechaRyRy) [inline library/src/main/java/com/novoda/storagepathfinder/ExternalFileDirectoryInspector.java]: I'd remove the comment, it doesn't add anything we didn't already know.

File: library/src/test/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspectorTest.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    private static final StoragePath PRIMARY_APPLICATION_PATH = DeviceStoragePath.create("/primary-application-path", PRIMARY);
+
+    private static final StoragePath SECONDARY_BASE_PATH_1 = DeviceStoragePath.create("/secondary-base-path-1", SECONDARY);
+    private static final StoragePath SECONDARY_BASE_PATH_2 = DeviceStoragePath.create("/secondary-base-path-2", SECONDARY);
+    private static final StoragePath SECONDARY_APPLICATION_PATH_1 = DeviceStoragePath.create("/secondary-application-path-1", SECONDARY);
+
+    private Context context = mock(Context.class);
+    private AndroidSystem androidSystem = mock(AndroidSystem.class);
+    private FileSystem filesystem = mock(FileSystem.class);
+    private DeviceFeatures devicesFeatures = mock(DeviceFeatures.class);
```
Comment (MechaRyRy) [inline library/src/test/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspectorTest.java]: All of the above mocks can be `final`.

File: library/src/test/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspectorTest.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+
+    private CommonDirectories commonDirectories = new CommonDirectories() {
```
Comment (MechaRyRy) [inline library/src/test/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspectorTest.java]: final?

File: library/src/test/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspectorTest.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+    @Test
+    public void canGetThePrimaryStorageBasePath() {
+        StoragePath basePath = storageInspector.getPrimaryStorageBasePath();
+
+        assertThat(basePath).isEqualTo(PRIMARY_BASE_PATH);
+    }
+
+    @Test
+    public void canGetThePrimaryStorageApplicationPath() {
```
Comment (MechaRyRy) [inline library/src/test/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspectorTest.java]: Can we agree on a naming structure for the tests? 😂 I dislike this 👅 

Would rather `returnsPrimaryStorageApplicationPath`

File: library/src/test/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspectorTest.java [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+        StoragePath basePath = storageInspector.getPrimaryStorageApplicationPath();
+
+        assertThat(basePath).isEqualTo(PRIMARY_APPLICATION_PATH);
+    }
+
+    @Test
+    public void canGetASecondaryStorageBasePathFromTheExternalDirectoryInspector() {
```
Comment (MechaRyRy) [inline library/src/test/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspectorTest.java]: This and the tests below seem to be leaking the implementation, I don't know anything about these inspectors from this classes point of view. Should we be writing these as a blackbox?

File: library/src/test/java/com/novoda/storagepathfinder/EnvironmentVariableStorageInspectorTest.java [hunk 15]
```diff
@@ -1,39 +1,53 @@
...
 
-        List<DeviceStorageRoot> secondaryDeviceStorageRoots = storageInspector.getSecondaryDeviceStorageRoots();
+        List<StoragePath> secondaryDeviceStorageRoots = storageInspector.getSecondaryDeviceStorageBasePaths();
 
-        assertThat(secondaryDeviceStorageRoots).isEmpty();
+        assertThat(secondaryDeviceStorageRoots.isEmpty()).isTrue();
+    }
+
+    @Test
+    public void returnsNoSecondaryStorageApplicationPathsWhenEnvironmentVariableIsAvailable() {
```
Comment (MechaRyRy) [inline library/src/test/java/com/novoda/storagepathfinder/EnvironmentVariableStorageInspectorTest.java]: Can we break up the `_when` so it's a bit easier to read.

**Review by MechaRyRy — CHANGES_REQUESTED** [review]
Good start 😄 just a few comments that need addressing.

File: library/build.gradle (hunk 4 above)
Comment (Hutch4) [reply library/build.gradle]: Oh, i've never used `truth`, didn't know about it, I can change this I guess. 
Though if `truth` is 0.41, is this not in Beta?

File: library/src/main/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspector.java (hunk 6 above)
Comment (Hutch4) [reply library/src/main/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspector.java]: Nope. 
If you pass a capacity to an ArrayList it sets the initial capacity, and if you add more elements it will then do some extra internal restructuring to increase the size. 
I will remove the 2 anyway though, as 95% of the time this will be 1.

File: library/src/main/java/com/novoda/storagepathfinder/DeviceStoragePath.java (hunk 9 above)
Comment (Hutch4) [reply library/src/main/java/com/novoda/storagepathfinder/DeviceStoragePath.java]: It is a path?  `path.pathAsFile()` doesn't make sense to me. `path.asFile()` reads more clearly

File: library/src/test/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspectorTest.java (hunk 13 above)
Comment (Hutch4) [reply library/src/test/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspectorTest.java]: Returns makes more sense 👍

File: library/src/main/java/com/novoda/storagepathfinder/DeviceStoragePath.java (hunk 9 above)
Comment (Hutch4) [reply library/src/main/java/com/novoda/storagepathfinder/DeviceStoragePath.java]: `getPathAsFile()` and `getPathAsString()` make sense. 

I know you are not a fan of `getX` but its common Java convention so for a OS library I'd suggest we use it

File: library/src/main/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspector.java (hunk 6 above)
Comment (MechaRyRy) [reply library/src/main/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspector.java]: Very true, doing a restructure is still bad 😄

File: demo/build.gradle [hunk 16]
```diff
@@ -14,8 +14,8 @@ android {
 
     lintOptions {
         lintConfig teamPropsFile('static-analysis/lint-config.xml')
-        abortOnError true
-        warningsAsErrors true
+        abortOnError false
+        warningsAsErrors false
```
Comment (MechaRyRy) [inline demo/build.gradle:18]: do you need to specify these? Wouldn't they be false by default?

File: library/src/test/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspectorTest.java [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+        givenFileSystemHasPath(SECONDARY_BASE_PATH_1);
+
+        List<StoragePath> storagePaths = storageInspector.getSecondaryStorageBasePath();
+
+        assertThat(storagePaths.size()).isEqualTo(1);
+        assertThat(storagePaths.get(0)).isEqualTo(SECONDARY_BASE_PATH_1);
+    }
+
+    @Test
+    public void noSecondaryStorageApplicationPathFromTheEnvironmentVariableInspector() {
```
Comment (MechaRyRy) [inline library/src/test/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspectorTest.java]: `returnsNo` like you did for the others

File: library/src/test/java/com/novoda/storagepathfinder/ExternalFileDirectoryInspectorTest.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+    private ExternalFileDirectoryInspector inspector;
+
+    @Before
+    public void setUp() {
+        inspector = new ExternalFileDirectoryInspector(context, deviceFeatures, PRIMARY_STORAGE_BASE_PATH);
+    }
+
+    @Test
+    public void canGetSecondaryStorageBasePathWhenFileDirectoriesAreAvailable() {
```
Comment (MechaRyRy) [inline library/src/test/java/com/novoda/storagepathfinder/ExternalFileDirectoryInspectorTest.java]: Not using `returns` in these tests.

**Review by MechaRyRy — CHANGES_REQUESTED** [review]
Missed a couple of test names

File: demo/build.gradle (hunk 16 above)
Comment (Hutch4) [reply demo/build.gradle:18]: Maybe?It might be nice to be explicit though?  
Docs don't say if it is 🤷‍♂️ 
http://google.github.io/android-gradle-dsl/current/com.android.build.gradle.internal.dsl.LintOptions.html#com.android.build.gradle.internal.dsl.LintOptions

Comment (MechaRyRy) [reply demo/build.gradle:18]: 🤷‍♂️ keep them then 😄

**Review by MechaRyRy — APPROVED** [review]
